src/core/reaction_ensemble.cpp

```diff
@@ -0,0 +1,2061 @@
+		check_is_successfull=ES_ERROR;
+	}
+	#ifdef ELECTROSTATICS
+	//check for the existence of default charges for all types that take part in the reactions
+	for (int i=0; i<current_reaction_system.nr_different_types; i++) {
+		if(current_reaction_system.charges_of_types[current_reaction_system.type_index[i]]==invalid_charge) {
+			printf("Forgot to assign default charges for some types\n");
+			check_is_successfull=ES_ERROR;
+		}
+	}
+	#endif
+	return check_is_successfull;
+}
+
+int free_reaction_ensemble(){
+	//needs to be called at the end of the simulation
+	for(int single_reaction_i=0;single_reaction_i<current_reaction_system.nr_single_reactions;single_reaction_i++){
+		//free reactant types and coefficients
+		free(current_reaction_system.reactions[single_reaction_i]->reactant_types);
+		free(current_reaction_system.reactions[single_reaction_i]->reactant_coefficients);	
+		//free product types and coefficients
+		free(current_reaction_system.reactions[single_reaction_i]->product_types);
+		free(current_reaction_system.reactions[single_reaction_i]->product_coefficients);
+		free(current_reaction_system.reactions[single_reaction_i]);	
+	}
+	free(current_reaction_system.reactions);
+	free(current_reaction_system.type_index);
+	free(current_reaction_system.charges_of_types);
+
+	return 0;
+}
+
+
+//boring helper functions
+
+void set_cuboid_reaction_ensemble_volume(){
+	if (current_reaction_system.volume <0)
+		current_reaction_system.volume=box_l[0]*box_l[1]*box_l[2];
+}
+
+float factorial_Ni0_divided_by_factorial_Ni0_plus_nu_i(int Ni0, int nu_i) {
+	float value=1;
+	if (nu_i == 0) {
+		value=1.0;
+	} else {
+		value=1.0;
+		if (nu_i > 0) {
+			for(int i=1;i <= nu_i;i++) {
+				value=value*1.0/(Ni0+i);
+			}
+		} else {
+			int abs_nu_i =-1.0*nu_i;
+			for(int i=0;i < abs_nu_i;i++) {
+				value= value*(Ni0-i);
+			}
+		}
+	}
+	return value;
+}
+
+
+bool all_reactant_particles_exist(int reaction_id) {
+	bool enough_particles=true;
+	for(int i=0;i<current_reaction_system.reactions[reaction_id]->len_reactant_types;i++){
+		int current_number;
+		number_of_particles_with_type(current_reaction_system.reactions[reaction_id]->reactant_types[i], &current_number);
+		if(current_number<current_reaction_system.reactions[reaction_id]->reactant_coefficients[i]){
+			enough_particles=false;
+			break;
+		}
+	}
+	return enough_particles;
+}
+
+double calculate_current_potential_energy_of_system(int unimportant_int){
+	//calculate potential energy
+	if (total_energy.init_status == 0) {
+		init_energies(&total_energy);
+		master_energy_calc();
+	}
+  	int num_energies=total_energy.data.n;
+	double kinetic_energy =total_energy.data.e[0];
+	double sum_all_energies=0;
+	for(int i=0;i<num_energies;i++){
+		sum_all_energies+= total_energy.data.e[i];
+	}
+	for (int i = 0; i < n_external_potentials; i++) {
```

Comment:
What is done here is to sum up all individual energy contributions. This is not done in master_energy_calc(). I agree with you that this could be done in a function in energy.cpp.
For now this kind of task always happened in the interface e.g. in analyze.pyx
```
        for i in range(c_analyze.total_energy.data.n):
            tmp += c_analyze.total_energy.data.e[i]

        e["total"] = tmp
```
Should I change this?